Update e2e test label definitions for FeatureGate and NodeConformance (#8923)
kubernetes/community: contributors/devel/sig-testing/e2e-tests.md

monitored closely in CI. `[Flaky]` tests are by default not run, unless a
`focus` or `skip` argument is explicitly given.

>   - `[Feature:.+]`: If a test validates functionality that may only work

Reviewer: do we feel this includes windows specific features? I think it may be good to callout our relationship to windows features and features that node  drives that are not and never will be windows

Author: I'd like to handle Windows separately. This page already has [LinuxOnly] and I do not think there is an established and recognized pattern for WindowsOnly

outside the minimal conformant installation of Kubernetes, e.g. when
specific node capabilities are enabled, having a certain addon available,
like loadbalancer integration, or when test depends on the
functionality of underlying components like container runtime, and thus may
need to be skipped on certain environments, it receives a `[Feature:.+]`
label, e.g. `[Feature:AppArmor]`.
`[Feature:.+]` label should not be confused with `[FeatureGate:]` label.
Tests marked with `[Feature:.+]` must be run in test jobs
that preconfigure the required environment. These jobs may run normal tests
as well as tests marked with this `[Feature:.+]` label.

  - `[FeatureGate:.+]`: If a test only works when a certain feature gate is
enabled it receives a `[FeatureGate:.+]` label. `[FeatureGate:.+]` tests
automatically get marked with the status of this feature gate: `[Alpha]` or
`[Beta]`. If a feature gate is GA, it has no level string. If a feature gate is
disabled by default, it also receives the `[Feature:OffByDefault]` label.
This label helps to skip tests that would not work on a specific Kubernetes
cluster that has a certain feature gate disabled. This label has to be

> removed when the feature gate value removed.

Reviewer 2: ```suggestion
removed when the feature gate definition gets removed.
```

  - `[Feature:OffByDefault]`: This label is added to tests that depend on a
feature gate that is disabled by default. It allows these tests to be easily

> skipped in environments with the default kubernetes deployment.

Reviewer 2: ```suggestion
```

This gets mentioned above and added below under "automatically added", which is IMHO the right place. This bullet item here seems redundant to me.

  - `[NodeConformance]`: Node-level tests that validating behavior that doesn't
depend on specific Node capabilities being present, hardware, or feature set
of a dependency (like a container runtime), must be labeled as `[NodeConformance]`.
For the ease of test querying, each node-level test that is not testing alpha feature
(marked as `[FeatureGate:Foo][Alpha]`) should be either `NodeConformance`

> or `Feature` (`[Feature:OffByDefault]` is predefined tag that also not compatible with `NodeConformance`).

Reviewer 2: ```suggestion
or `Feature` (`[Feature:OffByDefault]` is predefined tag that also not compatible with `NodeConformance`).
`NodeConformance` has been superseded by
`Feature` and will be removed.
```

monitored closely in CI. `[Flaky]` tests are by default not run, unless a
`focus` or `skip` argument is explicitly given.

  - `[Feature:.+]`: If a test validates functionality that may only work
outside the minimal conformant installation of Kubernetes, e.g. when

> specific node capabilities are enabled, having a certain addons available,

Reviewer 2: ```suggestion
specific node capabilities are enabled, having a certain addon available,
```

Author: applied

like loadbalancer integration, or when test depends on the
functionality of underlying components like container runtime, and thus may
need to be skipped on certain environments, it receives a `[Feature:.+]`
label, e.g. `[Feature:AppArmor]`. This label can also be applied to signify that

> test validates non-core functionality, like `[Feature:Ingress]`. 

Reviewer 2: ```suggestion
label, e.g. `[Feature:AppArmor]`.
```

`Feature` should not be used to categorize tests. Normal labels can be used for that.

Author: applied

> `[Feature:.+]` label is opposite of `[NodeConformance]` for node-specific tests. 

Reviewer 2: I don't understand what "is opposite of" means here.

Author: removed

`[Feature:.+]` label should not be confused with `[FeatureGate:]` label.

> Tests marked with `[Feature:.+]` must have corresponding test lanes declared

Reviewer 2: "test lanes" is a term that is only used by SIG node, as far as I know. Let's avoid introducing it here for the whole project.

 What we talk about in SIG Testing/Infra is "jobs". Let's use that terminology?

Author: removed

> that preconfigure the required environment. These test lanes may run `[NodeConformance]`

Reviewer 2: At this point, `[NodeConformance]` hasn't been introduced yet. Also, not all jobs are node jobs. Let's generalize?

```suggestion
that preconfigure the required environment. These jobs may run normal tests
```

tests as well as tests marked with this `[Feature:.+]` label.

  - `[FeatureGate:.+]`: If a test only works when a certain feature gate is
enabled it receives a `[FeatureGate:.+]` label. `[FeatureGate:.+]` tests

> must also be marked with the status of this feature gate: `[Alpha]` or

Reviewer 2: ```suggestion
automatically get marked with the status of this feature gate: `[Alpha]` or
```

`[Beta]`. If a feature gate is GA, it has no level string. If a feature gate is
disabled by default, it also receives the `[Feature:OffByDefault]` label.
This label helps to skip tests that should not work on a specific k8s

> distributive that has a certain feature gate disabled. This label has to be

Reviewer 2: ```suggestion
This label helps to skip tests that would not work on a specific Kubernetes
cluster that has a certain feature gate disabled. This label has to be
```

> removed when the feature gate value is "locked" or removed.

Reviewer 2: ```suggestion
removed when the feature gate value removed.
```

Thanks for version emulation, features aren't really "locked" anymore until much later. Let's keep it simple and just say "when removed", which conveniently (and intentionally!) is also when the E2E test code will stop to compile...

  - `[Feature:OffByDefault]`: This label is added to tests that depend on a
feature gate that is disabled by default. It allows these tests to be easily
skipped in environments with the default kubernetes deployment.

  - `[NodeConformance]`: Node-level tests that validating behavior that doesn't
depend on specific Node capabilities being present, hardware, or feature set

> of a dependency (like a container runtime), must be labeled as `[NodeConformance]`.

Reviewer 2: Let me be blunt: is `[NodeConformance]` really a thing nowadays? Is SIG Node still working on maintaining this label consistently and does anyone use it? It has nothing to do with "conformance" in the Kubernetes and CNCF sense.

In other words, can we remove it?

cc @BenTheElder @dims

Author: it is not used as conformance. It was useful in the past when Feature was not applied consistently and it was a way to query tests that must run in all environments. With Feature being consistently applied, I agree it is not needed.

Let's remove it!

cc @haircommander

Author: (perhaps we can land this PR first to reflect the "today's use of it", than separate work to remove it)

For the ease of test querying, each node-level test that is not testing alpha feature
(marked as `[FeatureGate:Foo][Alpha]`) is supposed to be either `NodeConformance`

> or `Feature` (`[Feature:OffByDefault]` is predefined tag that also not compatible with `NodeConformance`).

Reviewer 2: If we have such a requirement, then we should enforce it instead of going for a fuzzy "supposed to be".

    have to be annotated with a `[Feature]` tag. This restriction will be lifted
    once migration of jobs to `--filter-label` is completed.

### Decision making tree for the test labels

> - Is test validate functionality that is controlled by a Feature Gate? Apply `[FeatureGate:Foo]` and it will automatically get `[Alpha]` or `[Beta]` and `[Feature:OffByDefault]` if applicable.

Reviewer 2: ```suggestion
- Is test validate functionality that is controlled by a Feature Gate? Add `[FeatureGate:Foo]` with the E2E framework helper functions and it will automatically get `[Alpha]` or `[Beta]` and `[Feature:OffByDefault]` if applicable.
```

- Is the test validating a Core API that is enabled by default, GA, and works on any environment? Consider promotion to `[Conformance]`. See more at [conformance-tests.md](../sig-architecture/conformance-tests.md).

> - Is test only works when underlying container runtime has a specific feature enabled or specific node configuration is set? Apply `[Feature:Foo]` to describe this feature or configuration.

Reviewer 2: ```suggestion
- Is test only works when underlying container runtime has a specific feature enabled or specific node configuration is set? Apply `[Feature:Foo]` to declare that dependency.
  In the definition of the feature, describe it in
  enough detail so that others will know how to configure a cluster which can run the test.
```

    have to be annotated with a `[Feature]` tag. This restriction will be lifted
    once migration of jobs to `--filter-label` is completed.

### Decision making tree for the test labels

- Is test validate functionality that is controlled by a Feature Gate? Apply `[FeatureGate:Foo]` and it will automatically get `[Alpha]` or `[Beta]` and `[Feature:OffByDefault]` if applicable.

> - Is test validate Core API that is enabled by default, GA, and works on any environment? Apply `[Conformance]`. See more at [conformance-tests.md](../sig-architecture/conformance-tests.md).

Reviewer 2: ```suggestion
- Is the test validating a Core API that is enabled by default, GA, and works on any environment? Consider promotion to `[Conformance]`. See more at [conformance-tests.md](../sig-architecture/conformance-tests.md).
```

Not all such tests need to be promoted.
